File: CVS_ClassifierSample_ObjC/CVS_ClassifierSample/BubbleLabelLayer.m
// Copyright (c) Microsoft Corporation.  All rights reserved.

#import <UIKit/UIKit.h>
#import "BubbleLabelLayer.h"

@interface BubbleLabelLayer ()

@property (assign, nonatomic) CGSize preferredSize;

@end

@implementation BubbleLabelLayer

- (void)setPreferredSize:(CGSize)preferredSize {
    _preferredSize = preferredSize;
    [self setNeedsLayout];
}

- (void)updatePreferredSizeWithMaxWidth:(CGFloat)maxWidth {
    NSString *string = (NSString *)self.string;
    if (!string) {
        self.preferredSize = CGSizeZero;
        return;
    }
    
    UIFont *font = (UIFont *)self.font;
    if (!font) {
        self.preferredSize = CGSizeZero;
        return;
    }

    NSAttributedString *nsString = [[NSAttributedString alloc] initWithString:string attributes:@{NSFontAttributeName: font}];
    CGRect fontBounds = [nsString boundingRectWithSize:CGSizeMake(maxWidth, CGFLOAT_MAX)
                                            options:NSStringDrawingUsesLineFragmentOrigin
                                            context:nil];
    fontBounds.size.height += fabs(font.descender);
    self.preferredSize = fontBounds.size;
}

- (CGSize)preferredFrameSize {
    return self.preferredSize;
}

@end

File: CVS_ClassifierSample_Swift/CVS_ClassifierSample/BubbleLabelLayer.swift
// Copyright (c) Microsoft Corporation.  All rights reserved.

import UIKit

class BubbleLabelLayer : CATextLayer {
    
    private var preferedSize = CGSize.zero {
        didSet {
            needsLayout()
        }
    }
    
    func updatePreferredSize(maxWidth: CGFloat) {
        
        guard let string = self.string as? String else {
            //print("Trying to update label size without string")
            preferedSize = CGSize.zero
            return
        }
        
        guard let font = self.font as? UIFont else {
            print("Trying to update label size without font")
            preferedSize = CGSize.zero
            return
        }
        
        let nsString = NSAttributedString(string: string, attributes: [ .font : font ])
        var fontBounds = nsString.boundingRect(with: CGSize(width: maxWidth,
                                                            height: CGFloat.greatestFiniteMagnitude),
                                               options: [.usesLineFragmentOrigin],
                                               context: nil)
        fontBounds.size.height += abs(font.descender)
        preferedSize = fontBounds.size
    }
    
    override func preferredFrameSize() -> CGSize {
        return preferedSize
    }
}
